core/src/main/scala/pureconfig/ConfigSource.scala

```diff
@@ -0,0 +1,299 @@
+   * @param cur the cursor to be provided by this source
+   * @return a `ConfigObjectSource` providing the given cursor.
+   */
+  def fromCursor(cur: FluentConfigCursor): ConfigObjectSource =
+    new ConfigObjectSource(cur.asObjectCursor.right.map(_.value.toConfig))
+}
+
+object ConfigSource {
+
+  /**
+   * A config source for the default loading process in Typesafe Config. Typesafe Config stacks
+   * `reference.conf` resources provided by libraries, application configs (by default
+   * `application.conf` in resources) and system property overrides, resolves them and merges them
+   * into a single config.
+   */
+  val default = ConfigObjectSource(ConfigFactoryWrapper.load())
+
+  /**
+   * A config source that always provides empty configs.
+   */
+  val empty = ConfigObjectSource(Right(ConfigFactory.empty))
+
+  /**
+   * A config source for Java system properties.
+   */
+  val systemProperties = ConfigObjectSource(ConfigFactoryWrapper.systemProperties())
+
+  /**
+   * A config source for the default reference config in Typesafe Config (`reference.conf`
+   * resources provided by libraries).
+   *
+   * As required by
+   * [[https://github.com/lightbend/config/blob/master/HOCON.md#conventional-configuration-files-for-jvm-apps the HOCON spec]],
+   * the default reference files are pre-emptively resolved - substitutions in the reference config
+   * aren't affected by application configs.
+   */
+  val defaultReference = ConfigObjectSource(ConfigFactoryWrapper.defaultReference())
+
+  /**
+   * A config source for the default reference config in Typesafe Config (`reference.conf`
+   * resources provided by libraries) before being resolved. This can be used as an alternative
+   * to `defaultReference` for use cases that require `reference.conf` to depend on
+   * `application.conf`.
+   */
+  val defaultReferenceUnresolved = resources("reference.conf")
+
+  /**
+   * A config source for the default application config in Typesafe Config (by default
+   * `application.conf` in resources).
+   */
+  val defaultApplication = ConfigObjectSource(ConfigFactoryWrapper.defaultApplication())
+
+  /**
+   * A config source for the default loading process in Typesafe Config with a custom application
+   * config source. Typesafe Config stacks `reference.conf` resources provided by libraries, the
+   * given file and system property overrides, resolves them and merges them into a single config.
+   *
+   * This method is provided here to support use cases that previously depended on
+   * `ConfigFactory.load(config)`. Creating a custom source by merging the layers manually is
+   * usually recommended as it makes the config priorities more transparent.
+   *
+   * @param appSource the source providing the application config
+   * @return a `ConfigObjectSource` for the default loading process in Typesafe Config with a
+   *         custom application config source.
+   */
+  def applicationConf(appSource: ConfigObjectSource): ConfigObjectSource =
```

Comment:
We need a better name - this actually includes *everything but* `application.conf` :smile:. How about `withDefaultReferenceAndOverrides`?

Reply:
The goal of this method is to specify your application config (not `application.conf`) in the argument, but I agree that I don't exactly excel at naming things 🙃 `withDefaultReferenceAndOverrides` is certainly more descriptive than mine, but when I name methods as `withX` I expect to provide `X` as the argument (like in `withFallback`), while here it goes the other way around - we wrap something else in `X`.

@jcazevedo, @leifwickland, do you have any suggestion?

Reply:
You didn't ask if I had a good suggestion, so I'll offer my bad ones:
- `sandwichWithDefaults`
- `wrappedInDefaults`
- `addPreAndPostDefaults`
- `addDefaultReferenceAndOverrides`

If it weren't for the fact that we need this method to tell a simpler upgrade story, I'd say we should omit it since nobody will guess what it does without reading the source.

Reply:
Well, it's not off the table to delete this method and recommend users to migrate to `defaultOverrides.withFallback(ConfigSource.file(path)).withFallback(defaultReference)`. We're already deprecating `loadConfigFromFiles` without an alternative and the fact that we're having trouble naming it and "nobody will guess what it does without reading the source" makes me think this should never be provided in the first place.

Reply:
I'm not sure that providing this method as the default suggestion in the deprecation notes is what most people expect. My guess is that most users would be surprised to learn that when they tell pureconfig to load a file, it's also looking at multiple other sources.

I'd lean toward not having this method, but since I've never wanted the pre-and-post-defaults behavior I'm definitely not the target market, and thus won't express a strong opinion.

Reply:
> I name methods as `withX` I expect to provide `X` as the argument (like in `withFallback`), while here it goes the other way around - we wrap something else in `X`.

Certainly, `withDefaultReferenceAndOverrides` would work better as a part of the fluent API:
```scala
ConfigSource.defaultApplication
  .withDefaultReferenceAndOverrides
```
We can also split that into two orthogonal methods:
```scala
ConfigSource.defaultApplication
  .withDefaultReference
  .withDefaultOverrides
```

> My guess is that most users would be surprised to learn that when they tell pureconfig to load a file, it's also looking at multiple other sources.

That speaks more about the lack of transparency in the old API (and the `ConfigFactory` API it's based on). I don't think it's a bad thing to get it all out in the open.

> I'd lean toward not having this method

I'm *for* having this method, or some variation of it. This is the default config loading pattern (as [mentioned](https://github.com/lightbend/config/blob/558c1e22f53c7f165daaf3e1e792399087d79f93/HOCON.md#conventional-configuration-files-for-jvm-apps) in the HOCON spec), so it makes sense to provide a cookie-cutter implementation for anyone switching over from other HOCON libraries.

Reply:
I don't like the idea of polluting the `ConfigSource` trait with Typesafe Config specific defaults, let's keep Typesafe's opinionated config stack at the edge of the API :)

If we are to add this: since the "full" default process is called `.default`, what do you think of naming this `defaultWithCustomAppConf`?

Reply:
> `defaultWithCustomAppConf`

This has the same problem that `applicationConf` has: it focuses more on what this function doesn't provide.

After this discussion, I'm starting to lean towards one of @leifwickland's suggestions for the function name. `sandwichWithDefaults` (or `defaultSandwich` to match other methods) is memorable and descriptive, albeit unorthodox.

Reply:
What if we remove this method and modify the signature of `ConfigSource.default` to become `def default(appSource: ConfigObjectSource = defaultApplication)`? It'd map to the `ConfigFactory.load` API and save us from an extra method that's only there to provide compatibility with the HOCON spec.

Reply:
@jcazevedo, that works, too. However, in my (limited) experience, default parameter values tend to be a pain where binary compatibility is concerned. We can achieve the same with overloading:
```scala
def default(appSource: ConfigObjectSource) = ???
def default(): ConfigSource = default(defaultApplication)
```

Reply:
> This has the same problem that `applicationConf` has: it focuses more on what this function doesn't provide.

While I agree that `applicationConf` is very unclear on what it does, I don't see any problem with `defaultWithCustomAppConf`. I don't want to specify in any way what this function actually provides, as its purpose (and the only reason this function exists) is purely to delegate the loading strategy to Typesafe Config. The source of truth of what it actually does should be Typesafe Config's documentation, as there's nothing here preventing them from changing how loading works in a later version (which may not involve sandwiching).

I'm equally fine with having a `default(appSource: ConfigObjectSource)` (even though I don't see how it's any different from `defaultWithCustomAppConf`), so I'll go with that one since everyone agrees with it 🙂

![](https://starecat.com/content/wp-content/uploads/actual-programming-vs-debating-30-minutes-on-how-to-name-a-variable-drake-meme.jpg)

Reply:
Done in 98c9032.